tool/update_version.dart

```diff
@@ -0,0 +1,156 @@
+
+  final process = await Process.start('./tool/pub_upgrade.sh', []);
+  process.stdout.asBroadcastStream().listen((event) {
+    print(utf8.decode(event));
+  });
+}
+
+String incrementVersion(String oldVersion) {
+  final semVer = RegExp('[0-9]+\.[0-9]\.[0-9]+').firstMatch(oldVersion)[0];
+
+  const devTag = '-dev';
+  final isDevVersion = oldVersion.contains(devTag);
+  if (isDevVersion) {
+    return semVer;
+  }
+
+  final parts = semVer.split('.');
+
+  // Versions should have the form x.y.z
+  if (parts.length != 3) return null;
+
+  final patch = int.parse(parts.last);
+  final nextPatch = patch + 1;
+  return [parts[0], parts[1], nextPatch].join('.');
+}
+
+String versionFromPubspecFile(File pubspec) {
+  final lines = pubspec.readAsLinesSync();
+  for (final line in lines) {
+    if (line.startsWith(pubspecVersionPrefix)) {
+      return line.substring(pubspecVersionPrefix.length).trim();
+    }
+  }
+  return null;
+}
+
+void writeVersionToPubspec(File pubspec, String version) {
+  final lines = pubspec.readAsLinesSync();
+  final revisedLines = <String>[];
+  var currentSection = '';
+  final sectionRegExp = RegExp('([a-z]|_)+:');
+  for (var line in lines) {
+    if (line.startsWith(sectionRegExp)) {
+      // This is a top level section of the pubspec.
+      currentSection = sectionRegExp.firstMatch(line)[0];
+    }
+    if (editablePubspecSections.contains(currentSection)) {
+      if (line.startsWith(pubspecVersionPrefix)) {
+        line = [
+          line.substring(
+            0,
+            line.indexOf(pubspecVersionPrefix) + pubspecVersionPrefix.length,
+          ),
+          ' $version',
+        ].join();
+      } else {
+        for (final prefix in devToolsDependencyPrefixes) {
+          if (line.contains(prefix)) {
+            line = [
+              line.substring(0, line.indexOf(prefix) + prefix.length),
+              version,
+            ].join();
+            break;
+          }
+        }
+      }
+    }
+    revisedLines.add(line);
+  }
+  final content = revisedLines.joinWithNewLine();
+  pubspec.writeAsStringSync(content);
+}
+
+void writeVersionToVersionFile(File versionFile, String version) {
+  const prefix = 'const String version = ';
+  final lines = versionFile.readAsLinesSync();
+  final revisedLines = <String>[];
+  for (var line in lines) {
+    if (line.startsWith(prefix)) {
+      line = [prefix, '\'$version\';'].join();
+    }
+    revisedLines.add(line);
+  }
+  versionFile.writeAsStringSync(revisedLines.joinWithNewLine());
+}
+
+void writeVersionToChangelog(File changelog, String version) {
+  final lines = changelog.readAsLinesSync();
+  final versionString = '## $version';
+  if (lines.first.endsWith(versionString)) {
+    print('Changelog already has an entry for version $version');
+    return;
+  }
+  changelog.writeAsString([
+    '$versionString\n',
+    'TODO: update changelog',
+    ...lines,
+  ].joinWithNewLine());
```

Comment:
seems like this adds an extra trailing newline?

High level, consider https://github.com/dart-lang/pubspec_parse
rather than using regexp.

Reply:
it does. We want a new line at the end of the file.